### PR TITLE
Cubedsphere

### PR DESCRIPTION
**Pull request by colinjcotter**
This code implements a cubed sphere mesh, with gnonomic projection. I also left some code there for the dmplex refinement approach, which currently causes dmplex to segfault, since dmplex doesn't expect embedded quads in 3D.

### CODE REVIEW

File: firedrake/utility_meshes.py [hunk 1]
```diff
@@ -572,3 +573,222 @@ def UnitIcosahedralSphereMesh(refinement_level=0, reorder=None):
...
                                  reorder=reorder)
+
+
+def _cubed_vertices_and_faces(radius, refinement_level):
+    """
+    Generate vertex and face lists for cubed sphere
+    """
+    #We build the mesh out of 6 panels of the cube
+    #this allows to build the gnonomic cube transformation
+    #which is defined separately for each panel
```
Comment (kynan) [inline firedrake/utility_meshes.py]: Strict PEP8 would require a space after the leading `#`

File: tests/regression/test_embedded_sphere.py [hunk 2]
```diff
@@ -17,7 +16,6 @@ def integrate_sphere_area(radius=1, refinement=2):
     return np.abs(assemble(f * dx) - exact) / exact
 
 
-@longtest
```
Comment (kynan) [inline tests/regression/test_embedded_sphere.py]: Is this test now shorter?

**Comment by kynan** [comment]
Some commit squashing would be great (in particular squashing all the lint fixes with their respective origin commit).

File: tests/regression/test_embedded_sphere.py (hunk 2 above)
Comment (miklos1) [inline tests/regression/test_embedded_sphere.py]: Has this test ever been "long"? Commit message says that the 8 tests run in 11 seconds.

File: firedrake/utility_meshes.py [hunk 3]
```diff
@@ -568,3 +569,254 @@ def UnitIcosahedralSphereMesh(refinement_level=0, reorder=None):
...
+        for i in range(refinement_level):
+            plex = plex.refine()
+
+        #rescale points to the sphere
+        #this is not the same as the gnonomic transformation
+        vStart, vEnd = plex.getDepthStratum(0)
+        nvertices = vEnd - vStart
+        coords = plex.getCoordinatesLocal().array.reshape(nvertices, 3)
+        scale = (radius / np.linalg.norm(coords, axis=1)).reshape(-1, 1)
+        coords *= scale
```
Comment (dorugeber) [inline firedrake/utility_meshes.py]: Did this get fixed?  If not, what needs to be done here?

**Comment by dorugeber** [comment]
Done some commit-squashing, but breaking apart the two large linting commits just to squash them would seem an awful waste of time.

File: firedrake/utility_meshes.py (hunk 3 above)
Comment (colinjcotter) [inline firedrake/utility_meshes.py]: On 06/01/15 12:06, Doru Rathgeber wrote:
> In firedrake/utility_meshes.py
>> +
>> +        #rescale points to the sphere
>> +        #this is not the same as the gnonomic transformation
>> +        vStart, vEnd = plex.getDepthStratum(0)
>> +        nvertices = vEnd - vStart
>> +        coords = plex.getCoordinatesLocal().array.reshape(nvertices, 3)
>> +        scale = (radius / np.linalg.norm(coords, axis=1)).reshape(-1, 1)
>> +        coords *= scale
>
> Did this get fixed? If not, what needs to be done here?

It causes PETSc to segfault, I don't know why. I think this type of 
thing hasn't been done in PETSc before.

--cjc

**Comment by miklos1** [comment]
Okay, I have cleaned up whatever I dared to touch. (I still don't find the assembly of the panel numbering very nice, but it is hard to find a nice way to do that.)
So I approve the merge of the current state of this branch, however, due to the recent changes I made, an independent review would be great.

File: firedrake/utility_meshes.py [hunk 4]
```diff
@@ -572,3 +571,192 @@ def UnitIcosahedralSphereMesh(refinement_level=0, reorder=None):
...
+        plex = _from_cell_list(2, faces, vertices)
+        plex.setRefinementUniform(True)
+        for i in range(refinement_level):
+            plex = plex.refine()
+
+        # rescale points to the sphere
+        # this is not the same as the gnonomic transformation
+        coords = plex.getCoordinatesLocal().array.reshape(-1, 3)
+        scale = (radius / np.linalg.norm(coords, axis=1)).reshape(-1, 1)
+        coords *= scale
```
Comment (wence-) [inline firedrake/utility_meshes.py]: Given that the DMPlex refinement method of generating the cubed sphere doesn't actually produce the same mesh (in terms of locations of the vertices), I'm inclined just to drop this option completely.

Comment (wence-) [inline firedrake/utility_meshes.py]: Although, having said that, for MG on the cubed sphere, we'll need to be able to transform the coordinates of the refined mesh using said gnomonic projection, so maybe it's worthwhile writing the code to do that here.

File: tests/regression/test_embedded_sphere.py (hunk 2 above)
Comment (wence-) [inline tests/regression/test_embedded_sphere.py]: Many moons ago, yes, if running with debug petsc and `-malloc_debug`.

**Pull request merged**